querier.go

```diff
@@ -0,0 +1,365 @@
+	if err != nil {
+		q.metrics.valuesHistogram.WithLabelValues(connect.CodeOf(err).String()).Observe(latency.Seconds())
+		log.Printf("values(label=%s): failed to make request: %v\n", label, err)
+		return
+	}
+	q.metrics.valuesHistogram.WithLabelValues(grpcCodeOK).Observe(latency.Seconds())
+	log.Printf("values(label=%s): took %v and got %d results\n", label, latency, len(resp.Msg.LabelValues))
+}
+
+func (q *Querier) queryProfileTypes(ctx context.Context) {
+	queryStart := time.Now()
+	resp, err := q.client.ProfileTypes(ctx, connect.NewRequest(&queryv1alpha1.ProfileTypesRequest{}))
+	latency := time.Since(queryStart)
+	if err != nil {
+		q.metrics.profileTypesHistogram.WithLabelValues(connect.CodeOf(err).String()).Observe(latency.Seconds())
+		log.Println("profile types: failed to make request", err)
+		return
+	}
+	q.metrics.profileTypesHistogram.WithLabelValues(grpcCodeOK).Observe(latency.Seconds())
+	log.Printf("profile types: took %v and got %d types\n", latency, len(resp.Msg.Types))
+
+	if len(resp.Msg.Types) == 0 {
+		return
+	}
+
+	q.profileTypes = q.profileTypes[:0]
+	for _, pt := range resp.Msg.Types {
+		key := fmt.Sprintf("%s:%s:%s:%s:%s", pt.Name, pt.SampleType, pt.SampleUnit, pt.PeriodType, pt.PeriodUnit)
+		if pt.Delta {
+			key += ":delta"
+		}
+		q.profileTypes = append(q.profileTypes, key)
+	}
+}
+
+func (q *Querier) queryRange(ctx context.Context) {
+	if len(q.profileTypes) == 0 {
+		log.Println("range: no profile types to query")
+		return
+	}
+
+	profileType := q.profileTypes[len(q.profileTypes)/2]
+
+	for _, tr := range q.queryTimeRanges {
+		rangeEnd := time.Now()
+		rangeStart := rangeEnd.Add(-1 * tr)
+
+		queryStart := time.Now()
+		resp, err := q.client.QueryRange(ctx, connect.NewRequest(&queryv1alpha1.QueryRangeRequest{
+			Query: profileType,
+			Start: timestamppb.New(rangeStart),
+			End:   timestamppb.New(rangeEnd),
+		}))
+		latency := time.Since(queryStart)
+		if err != nil {
+			q.metrics.rangeHistogram.WithLabelValues(
+				connect.CodeOf(err).String(), tr.String(),
+			).Observe(latency.Seconds())
+			log.Printf("range(type=%s,over=%s): failed to make request: %v\n", profileType, tr, err)
+			continue
+		}
+
+		q.metrics.rangeHistogram.WithLabelValues(grpcCodeOK, tr.String()).Observe(latency.Seconds())
+		log.Printf(
+			"range(type=%s,over=%s): took %s and got %d series\n",
+			profileType, tr, latency, len(resp.Msg.Series),
+		)
+
+		for k := range q.series {
+			delete(q.series, k)
+		}
+		for _, series := range resp.Msg.Series {
+			if len(series.Samples) < 2 {
+				continue
+			}
+			lSet := labels.Labels{}
+			for _, l := range series.Labelset.Labels {
+				lSet = append(lSet, labels.Label{Name: l.Name, Value: l.Value})
+			}
+			sort.Sort(lSet)
+			q.series[profileType+lSet.String()] = timestampRange{
+				first: series.Samples[0].Timestamp.AsTime(),
+				last:  series.Samples[len(series.Samples)-1].Timestamp.AsTime(),
+			}
+		}
+	}
+}
+
+func (q *Querier) querySingle(ctx context.Context) {
```

Comment:
I don't think we need querySingle anymore. We have it available in Parca and Polar Signals Cloud, however, the UI is never using the API anymore. We should just remove it from parca-load.

Reply:
How does the UI display a single profile or get a profile to diff? I think this is still a very important query to have, as it highlighted deficiencies in object storage filtering recently and I think we want to add it to our OKRs next quarter.

Reply:
Instead of using this old API the UI now does a merge request where the start and end timestamps are the same. That way we only need one API in the future but still can request individual profiles for a timestamp. This old `querySingle` was mostly a shim for that.

Reply:
👍🏼 I'll open a separate issue for this too